helix-term/src/commands.rs

```diff
@@ -2980,7 +2982,38 @@ fn command_mode(cx: &mut Context) {
 
             // Handle typable commands
             if let Some(cmd) = cmd::TYPABLE_COMMAND_MAP.get(parts[0]) {
-                if let Err(e) = (cmd.fun)(cx, &parts[1..], event) {
+                let mut args: Vec<&str> = Vec::new();
+                let quote_pos: Vec<usize> = input
+                    .char_indices()
+                    .filter(|&(_idx, char)| char == '"')
+                    .map(|(idx, _char)| idx)
+                    .collect();
+
+                if quote_pos.len() < 2 {
+                    args = parts;
+                } else {
+                    args.append(
+                        &mut input[0..quote_pos[0]]
+                            .split_whitespace()
+                            .collect::<Vec<&str>>(),
+                    );
+                    for i in 0..quote_pos.len() / 2 {
```

Comment:
Seems to me that you're trying to simulate `.chunks(2)` (and I can't wait for `array_chunks(2)` to be stabilized).

Reply:
It might also be easier to scan over the string and use a state machine, that way we can also handle escaping and also handle single quotes: https://github.com/tmiasko/shell-words/blob/8f264f1f9ec463a523c751f7bb56a07371db2b53/src/lib.rs#L108-L211

I'd use a `Cow<'_, str>` so that we can avoid allocating parts without escapes.

Reply:
It's OK to do this in a follow-up though.

Reply:
Thanks for the info on chunks, i rewrote that part and i think the code is now a lot cleaner and easier to understand.

Reply:
I did it like this, because this way i can just use the str slices, no need to do extra allocations.
But in case it will be needed i can try to rewrite this using state machine.

Reply:
There's no need to allocate if you use a state machine and only track offsets:

`"` => start = pos
.... pos++
`"` => start.is_some() -> end = pos
words.push(input[start..end])

We only need to allocate if there's an escape character present. That's why I was suggesting `Cow<'_, str>`

Reply:
Ah yes, i understand it now, thanks for the explanation!

Reply:
I made the changes, and it now supports quotes, doublequotes, escaping, also open and hsplit/vsplit can now open multiple files at once.